File: c/circular-buffer/circular_buffer.c
#include <errno.h>
#include <stdint.h>
#include <stdlib.h>
#include "circular_buffer.h"

circular_buffer_t *new_circular_buffer(size_t capacity)
{
    circular_buffer_t *buffer = malloc(sizeof(circular_buffer_t));
    if (!buffer) {
        return NULL;
    }

    buffer->size = capacity;
    buffer->head = 0;
    buffer->tail = 0;
    buffer->count = 0;

    buffer->data = malloc(capacity * sizeof(buffer_value_t));
    if (!buffer->data) {
        free(buffer);
        buffer = NULL;
    }

    return buffer;
}

void clear_buffer(circular_buffer_t *buffer)
{
    buffer->head = 0;
    buffer->tail = 0;
    buffer->count = 0;
}

void delete_buffer(circular_buffer_t *buffer)
{
    free(buffer->data);
    free(buffer);
}

int16_t read(circular_buffer_t *buffer, buffer_value_t *value)
{
    if (buffer->count == 0) {
        errno = ENODATA;
        return EXIT_FAILURE;
    }

    buffer->count--;

    *value = buffer->data[buffer->head];
    buffer->head = (buffer->head + 1) % buffer->size;

    return EXIT_SUCCESS;
}

int16_t write(circular_buffer_t *buffer, buffer_value_t value)
{
    if (buffer->count == buffer->size) {
        errno = ENOBUFS;
        return EXIT_FAILURE;
    }

    buffer->count++;

    buffer->data[buffer->tail] = value;
    buffer->tail = (buffer->tail + 1) % buffer->size;

    return EXIT_SUCCESS;
}

int16_t overwrite(circular_buffer_t *buffer, buffer_value_t value)
{
    if (buffer->count == buffer->size) {
        buffer->head = (buffer->head + 1) % buffer->size;
    } else {
        buffer->count++;
    }

    buffer->data[buffer->tail] = value;
    buffer->tail = (buffer->tail + 1) % buffer->size;

    return EXIT_SUCCESS;
}

File: scala/circular-buffer/src/main/scala/CircularBuffer.scala
class EmptyBufferException() extends Exception {}

class FullBufferException() extends Exception {}

class CircularBuffer(var capacity: Int) {

  private val buffer     = Array.ofDim[Int](capacity)
  private var size       = capacity
  private var items      = 0
  private var readIndex  = 0
  private var writeIndex = 0

  def clear() = {
    items      = 0
    readIndex  = 0
    writeIndex = 0
  }

  def read(): Int = {
    if (items == 0) throw new EmptyBufferException()

    decreaseItems()

    val value = buffer(readIndex)
    shiftReadIndex()
    value
  }

  def write(value: Int) = {
    if (items == size) throw new FullBufferException()

    increaseItems()

    buffer.update(writeIndex, value)
    shiftWriteIndex()
  }

  def overwrite(value: Int) = {
    if (items == size) shiftReadIndex() else increaseItems()

    buffer.update(writeIndex, value)
    shiftWriteIndex()
  }

  private def decreaseItems() = items -= 1
  private def increaseItems() = items += 1

  private def shiftReadIndex()  = readIndex  = (readIndex  + 1) % size
  private def shiftWriteIndex() = writeIndex = (writeIndex + 1) % size

}
